Implement Leaderboard API Endpoint

anishprasad01 opened the pull request:
Implemented the Leaderboard API. Returns JSON of Pet Name, Score, and Owner Name of the Top Ten scoring pets in descending order

anishprasad01 committed:
Added LeaderboardController.cs

anishprasad01 committed:
Added LeaderboardController.cs

anishprasad01 committed:
Added query

anishprasad01 committed:
Added query

anishprasad01 committed:
Returns names of pets in desc. order in JSON

anishprasad01 committed:
Returns names of pets in desc. order in JSON

anishprasad01 committed:
returns pet names, scores, and owners names as json

anishprasad01 committed:
returns pet names, scores, and owners names as json

anishprasad01 committed:
Added link to Leaderboard API on the index page

anishprasad01 committed:
Added a condition to check for no Race data

anishprasad01 committed:
Added a condition to check for no Race data

anishprasad01 committed:
Added link to Leaderboard on Index.cshtml

anishprasad01 committed:
Deleted index.html (old file)

anishprasad01 committed:
Modified For loops to run only as long as elements are present in the lists. Added comments

anishprasad01 committed:
Added check for incongrous List lengths, and condensed code from 3 for loops to 1

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -0,0 +1,89 @@
+        //initialize connection to SQL Server
+        private readonly SqlManager sqlManager;
+
+        public LeaderboardController(SqlManager sqlManager) : base()
```
Specifying `base()` is not necessary.

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -0,0 +1,89 @@
+            //names of pets on leaderboard
+            List<string> TopTenPetNames = new List<string>();
+            //scores of pets on leaderboard
+            List<Int32> TopTenScores = new List<Int32>();
```
Use of boxed types like `Int32` isn't necessary in C#, `List<int>` works just fine.

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -0,0 +1,89 @@
+            //owner names of pets on leaderboard
+            List<string> TopTenOwnerNames = new List<string>();
+            //data structure to be returned as json
+            string[,] table = new string[3,10];
```
I would greatly prefer that a `List` of some new class were returned, instead of a 2D array of strings. This is much easier to work with, and preserves the data type (instead of converting everything into a string).

Chris-Johnston replied:
Also, `10` (aka the number of items to get should be defined as a constant. This way, should we decide to return more or less values, it's only changed once.

Chris-Johnston replied:
I'm also not a fan of returning a 3x10 array of null strings where there aren't any values.

anishprasad01 replied:
What would you prefer as a return value? Initialized to something?

Chris-Johnston replied:
> What would you prefer as a return value? Initialized to something?

Ideally it should return an empty list.

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -0,0 +1,89 @@
+                var cmd = conn.CreateCommand();
+
+                //to be optimized later, if time allows
+                cmd.CommandText = @"SELECT TOP (10) Pet.[Name] AS 'Pet Name', Race.Score, [User].Username AS 'Owner Name' 
```
Also should avoid hard-coding in the value of `10` here.

Chris-Johnston replied:
Query columns should not contain spaces: `Pet Name` `Owner Name` should instead be `PetName`, `OwnerName`

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -0,0 +1,89 @@
+                        //add each pet name/score/owner name to a new index in the list
+                        TopTenPetNames.Add(reader.GetString(0));
+                        TopTenScores.Add(reader.GetInt32(1));
+                        TopTenOwnerNames.Add(reader.GetString(2));
```
This is where it would be beneficial to create new objects of a new class type, and populate that into a single List, not three.

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -0,0 +1,89 @@
+                        TopTenOwnerNames.Count <= 0)
+                {
+                    //returns empty table full of nulls
+                    return Json(table);
```
Duplicate code of line 86 that doesn't do anything.

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -0,0 +1,89 @@
+                            (TopTenPetNames.Count != TopTenOwnerNames.Count))
+                {
+                    //returns empty table full of nulls
+                    return Json(table);
```
Duplicate code of line 86 that doesn't do anything.

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -0,0 +1,89 @@
+                    return Json(table);
+                }
+                //if there is data, add it to the table
+                else
```
You should check for the condition that there is data once, instead of checking for the absence of valid data twice.

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -0,0 +1,89 @@
+                        //add the pet names to index 0
+                        table[0, i] = TopTenPetNames[i];
+                        //add the scores to index 1
+                        table[1, i] = TopTenScores[i].ToString();
```
This is why I think a new class type should be created, instead of making everything a string.

Chris-Johnston commented on PetGame/Pages/Index.cshtml:
```diff
@@ -24,4 +24,7 @@
         <a href="login">Login Page</a>
     </li>
+    <li>
+        <a href="/api/leaderboard">Leaderboard Page</a>
```
Ideally we should not display the API endpoints directly to the end user. There should be some other page that calls this endpoint, gets the results, and displays it in a nicer-looking interface.

anishprasad01 replied:
I agree, this is just for testing until the Leaderboard Page is finished.

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -0,0 +1,89 @@
+        }
+
+        [HttpGet]
+        public IActionResult Get()
```
I think that this request should allow for optional parameters that allow for specifying the number of results, and the offset of results, so that the response could be paginated.

Chris-Johnston replied:
In addition, there should be an additional endpoint that allows for getting a single Leaderboard row (w/ names joined from the other tables) by the race ID.

Chris-Johnston reviewed (changes requested):
I think that this controller is OK, but I strongly believe that a RaceController needs to be implemented also to handle get/update/insert/delete of Races (this can be done in another PR). While it is more convenient to have a single endpoint that does everything, this underlying ability still needs to be there.

anishprasad01 committed:
Created new type: LeaderboardEntry.
Modified LeaderboardController to use the new type.
Removed hardcoded numbers.

anishprasad01 committed:
Moved LeaderboardEntry from nested in LeaderboardController to PetGame.Core Models

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -5,85 +5,63 @@
+    public class LeaderboardEntry
+    {
+        public string PetName { get; set; }
+        public int score { get; set; }
```
Please follow consistent naming patterns.

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -5,85 +5,63 @@
 
 namespace PetGame
 {
+    public class LeaderboardEntry
```
This should be in it's own file.

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -5,85 +5,63 @@
+        public int score { get; set; }
+        public string OwnerName { get; set; }
+
+        public LeaderboardEntry(string nPetName, int nScore, string nOwnerName)
```
This constructor is unnecessary, initialization can be done with the following:
```cs
var x = new LeaderboardEntry() { PetName = "aaa", Score = 44, OwnerName = "person" };
```

Also, why wouldn't this type include all of the fields from Race, including the time when the score was set and the various IDs?

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -5,85 +5,63 @@
 
         [HttpGet]
-        public IActionResult Get()
+        public IActionResult Get(int NumResults = 10)
```
How is this variable set? You don't specify attributes to indicate that it's part of the request body or not. In addition, this could use xmldoc comments (`///` comments)

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -5,85 +5,63 @@
-                cmd.CommandText = @"SELECT TOP (10) Pet.[Name] AS 'Pet Name', Race.Score, [User].Username AS 'Owner Name' 
-                                    FROM Pet, Race, [User] 
-                                    WHERE Race.PetId = Pet.PetId AND [User].UserId = Pet.UserId ORDER BY Score DESC;";
+                cmd.CommandText = @"SELECT TOP (" + NumResults + ") Pet.[Name] AS 'PetName', Race.Score, [User].Username AS 'OwnerName' " +
```
Instead of string concatenation, use parameters (`cmd.Parameters.AddWithValue("@NumResults");`). Currently this is potentially SQLi vulnerable.

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -5,85 +5,63 @@
-                        TopTenOwnerNames.Add(reader.GetString(2));
-                    }
-                }
+                        string PetName = reader.GetString(0);
```
Don't need to create new variables that are immediately discarded.

Chris-Johnston commented on PetGame.Core/Models/LeaderboardEntry.cs:
```diff
@@ -0,0 +1,20 @@
+    public class LeaderboardEntry
+    {
+        public string PetName { get; set; }
+        public int score { get; set; }
```
Inconsistent variable names.

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -0,0 +1,54 @@
+    [Route("api/[controller]")]
+    public class LeaderboardController : Controller
+    {
+        //initialize connection to SQL Server
```
this is a false statement

anishprasad01 committed:
Working on adding a new endpoint for getting a Leaderboard entry by raceid
working on adding a LeaderboardService.cs file

anishprasad01 committed:
Created LeaderboardRequestModel to add a request body to the Leaderboard GET.

Moved logic from LeaderboardController to LeaderboardService

anishprasad01 committed:
Made request body for GET optional

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -9,46 +7,59 @@ namespace PetGame
+        [HttpGet("{id}")]
+        public IActionResult Get(ulong id)
+        {
+            if (!(id < 0))
```
`ulong` is an unsigned type. It will never be less than 0.

Chris-Johnston commented on PetGame/Models/LeaderboardRequestModel.cs:
```diff
@@ -0,0 +1,8 @@
+{
+    public class LeaderboardRequestModel
+    {
+        public int NumRequests { get; set; }
```
These should both be optional, and unsigned integers. Make them both nullable.

Chris-Johnston commented on PetGame/Services/LeaderboardService.cs:
```diff
@@ -0,0 +1,86 @@
+            this.sqlManager = sqlManager;
+        }
+
+        public IEnumerable<LeaderboardEntry> GetLeaderboardEntries(int NumResults = 10)
```
This needs to support the offset.

Chris-Johnston commented on PetGame/Services/LeaderboardService.cs:
```diff
@@ -0,0 +1,86 @@
+                var cmd = conn.CreateCommand();
+
+                //to be optimized later, if time allows
+                cmd.CommandText = @"SELECT TOP (@Count) Pet.[Name] AS 'PetName', Race.Score, [User].Username AS 'OwnerName'
```
`TOP` should be replaced with `OFFSET FETCH`

Chris-Johnston commented on PetGame/Services/LeaderboardService.cs:
```diff
@@ -0,0 +1,86 @@
+                                    FROM Pet, Race, [User]
+                                    WHERE Race.PetId = Pet.PetId AND [User].UserId = Pet.UserId AND Race.RaceId = @RaceID;";
+
+                cmd.Parameters.Add("@RaceID", SqlDbType.BigInt).Value = Convert.ToInt64(RaceId);
```
this is not the correct way to insert a `ulong` parameter. Instead, use `$"{raceid}"`.

Chris-Johnston commented on PetGame/wwwroot/css/main.css:
```diff
@@ -0,0 +1,4 @@
+.body-container {
```
This file should not be included, because it is auto-generated from LESS. Remove it.

anishprasad01 committed:
Added Timestamp, PetId, and UserId to returned LeaderboardEntry

Modified LeaderboardService to handle new columns

anishprasad01 committed:
Implemented OFFSET-FETCH.

Added xmldoc comments

anishprasad01 committed:
Removed main.css

Removed <0 check on ulong

Changed LeaderboardRequestModel int to uint

anishprasad01 committed:
Inserts ulong into SQL query correctly

anishprasad01 committed:
Removed checks that offset and numrequests are > 0 since the type is now uint

anishprasad01 committed:
Changed Offset and NumRequests back to int because uint is unsupported by OFFSET-FETCH

anishprasad01 committed:
Fixed CodeFactor error

Chris-Johnston commented on PetGame.Core/Models/LeaderboardEntry.cs:
```diff
@@ -0,0 +1,19 @@
+    public class LeaderboardEntry
+    {
+        public int Score { get; set; }
+        public DateTime Timestamp { get; set; }
```
See #44 for how to add `JsonConverterAttribute`, `JsonPropertyAttribute`, and `JsonObjectAttribute` attributes.

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -0,0 +1,94 @@
+        /// OwnerName, OwnerId
+        /// </returns>
+        [HttpGet]
+        public IActionResult Get([FromBody] LeaderboardRequestModel Request)
```
This model can never be supplied with a `GET`, I'd add a `[HttpPost("options")]` route that allows for these options to be specified in the json body.

anishprasad01 replied:
Add, or replace?

Chris-Johnston replied:
Add

Chris-Johnston commented on PetGame/Controllers/LeaderboardController.cs:
```diff
@@ -0,0 +1,94 @@
+        {
+            //if no request body is specified, it will be null
+            //if null, creates a new request with default values
+            if (Request == null)
```
Rename `Request`, this is a commonly used name and may be confusing.

Chris-Johnston commented on PetGame/Models/LeaderboardRequestModel.cs:
```diff
@@ -0,0 +1,11 @@
+    public class LeaderboardRequestModel
+    {
+        public int Offset { get; set; }
+        public int NumRequests { get; set; }
```
Rename `NumRequests` to `NumItems`. You are only making a single request, not many. Name is misleading.

Chris-Johnston commented on PetGame/Models/LeaderboardRequestModel.cs:
```diff
@@ -0,0 +1,11 @@
+    /// <summary>
+    /// Represents the request body for the Leaderboard API
+    /// </summary>
+    public class LeaderboardRequestModel
```
Add json attributes

Chris-Johnston reviewed (changes requested):
Can you also include sample data for making requests in Postman?

Chris-Johnston commented:
Also, please resolve merge conflicts.

anishprasad01 committed:
Added JsonAttributes to LeaderboardEntry and LeaderboardRequestModel

Renamed Request in LeaderboardController

Renamed NumRequests to NumItems

Added HttpPost tag

Added more xmldoc comments

Chris-Johnston commented:
Resolve merge conflicts and we should be good to go.

Chris-Johnston reviewed (changes requested):
Oops, still missing example request data

anishprasad01 committed:
Added example api calls in comments

anishprasad01 committed:
Updated example request comments

anishprasad01 committed:
Merge branch 'master' into leaderboard-api
